witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs

```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
         next <- maybe (pure []) (getFrom . Just) (plNextId txList)
         return (plItems txList : next)
 
+data SpoilVariant
+    = WrongAddr
+    | WrongHeaderHash
+    | WrongProof
+    deriving (Show, Eq, Generic, Enum, Bounded)
+
+-- | Spoils a @'MerkleProof'@ by changing some of its internal nodes
+-- TODO: spoil better.
+spoilMerkleProof
+    :: Monad m
+    => MerkleProof a
+    -> PropertyM m (MerkleProof a)
+spoilMerkleProof mproof = do
+    newSig <- pick arbitrary
+    pure $ mproof { pnSig = newSig }
+
+-- | Spoils a @'FairCV'@ by changing some of its parts to random parts.
+-- TODO: remove copy-paste
+spoilFairCV
+    :: Monad m
+    => FairCV Unchecked
+    -> PropertyM m (FairCV Unchecked)
+spoilFairCV (FairCV cv) = pick arbitraryBoundedEnum >>= \case
+    WrongAddr -> do
+        oldAddr <- pick . elements $ M.keys cv
+        newAddr <- pick arbitrary
+        let subCv = fromMaybe (error "impossible") $
+                    M.lookup oldAddr cv
```

Comment:
@Martoon-00 Which pair? Not sure I got what you mean.

Reply:
```
(key, value) <- elements (M.toList cv)
```
instead of
```
key <- elements (M.keys) cv
let value = fromMaybe (error "impossible") $
                 M.lookup oldAddr cv 
```
There seems to be no performance overhead of the former comparing to the latter, `elements` converts map to the list inside anyway.